### ISSUE TITLE
How to internal storage the array

### ISSUE THREAD

**Issue by dotChris90** [Oceania2018/NumSharp#35]
This issue come up because of discussion in other issue. 

We should clear how to internal storage the NDArray data. 
Until now we discussed 2 strategies. And I want to use this issue to discuss pro and cons. 

Option 1 ) NDArray as an adapter 
This means we have a generic NDArray class< T > where T can be any array or jagged array. The storage is 100% the same like the generic T. So accessing an object arr[idx][jdx] would lead to the same like arr.Data[idx][jdx] and the methods would just extend existing .NET classes.

Option 2 ) NDArray as a total new Array type
This means we have for all types e.g. matrix, tensor etc. always just one 1D array. The NDArray must offer methods for indexing and so on by itself and every type is always a NDArray< T >.

**Comment by fdncred** [comment]
> @Oceania2018 Do I really understand you well? So you want store everything (1D,2D,3D,...ND) in a single array? The properties like Shape decide the dimension? Do not get me wrong but this will leads to some ... problems I think.
> 
> 1 ) our methods will get longer and not so well structured. Until now we can have "MethodName(NDArray< double >)" and "MethodName(NDArray<double[]>" to differ between vector and matrix. Since polymorphism we can have 2 different methods with the same name but different parameters. If our objects are always NDArray you can not make this but instead always have to do a huge "if else" structure. If method see it is vector do this, if matrix this. So this leads to less files but also increase the danger of "people have to work on 1 file at the same time".
> 
> 2 ) It is not totally OOP in my opinion. In OOP we say "This is a matrix and it has this methods and properties" and this is a vector with properties and methods. But here we say it is an array - It can be anything. That is dynamic interpreter style not compiler. It is python - not C#.
> 
> 3 ) Performance. Sorry I say but I am not sure if a huge array brings better performance. We should do some tests to find out best but I am very sure jagged arrays are faster than 1 huge array and you have to search the elements first with every access.
> 
> 4 ) Do you really, really want to rewrite all the operations and methods? It will be hard because on Stackoverflow you will find code examples with double[][] and so on but never a example with double[] for a matrix.
> 
> 5 ) Why you want to create your own Array? .NET world already has very fine and optimised ways for arrays. Python not so they developed from scratch. So we should always stay with this array type system.
> Thats why I suggested NDArray<double[][]>. I know at end it is a jagged array. So I know at end the corresponding type and most important we can keep NDArray as an adapter class and not a new fancy own class - .NET does not need this.
> 
> So please give me some reason why NDArray< double > matrix = new NDArray< double >().Array(...) is better than NDArray<double[][]> matrix = new NDArray<double[][]>().Array() ?
> 
> **I know the QuantStack do it and I find little bit weird** since I CAN NOT SEE FROM MY CODE WHAT Numeric TYPE I HAVE. Have a look again :
> 
> var matrix = new NDArray<double[][]>() **// I can see 100% it must be a matrix**
> var matrix = new NDArray< double >() **// Is it a vector, a matrix? no! it is something we do not know -.-**
> 
> So give me some arguments and pros. I do not want something like "because QuantStack do like this". I want sth like "better performance for Matrix Multiplication" because honestly all the points I listed at moment makes me feel not comfortable with the QuantStack solution.

Just quoting original thread.

In response, I could care less about the internal structure of arrays. My attempt with Bitmaps was to try  and emulate python's numpy array as close as possible (but I failed). In the end, my vote will be to store the data in a way that gives the best performance and the best intuitive ease of use. Performance is king but not at the sacrifice of having a useable API.

**Comment by fdncred** [comment]
Thoughts on this?
```cs
public static NDArray<int[][]> ARange(this NDArray<int[][]> np, int height, int width, int depth, int start = 0, int step = 1)
{
    np.Data = new int[height][][];
    int dataLen = width * height;
    for (int i = 0; i < dataLen; i++)
    {
        var x = i % width;
        var y = i / width;
        if (x == 0)
            np.Data[y] = new int[width][];
        np.Data[y][x] = new int[] { 1, 2, 3 };
    }

    return np;
}
```
```cs
[TestMethod]
public void ARangeArray()
{
    var np = new NDArray<int[][]>();
    np.ARange(2531, 2081, 3);

    var h = np.Data.Length;
    var w = np.Data[0].Length;
    for (int y = 0; y < h; y++)
    {
        for (int x = 0; x < w; x++)
        {
            var data = np.Data[y][x];
            Assert.IsTrue(Enumerable.SequenceEqual(data, new int[] { 1, 2, 3 }));
        }
    }
}
```
This allows me to do this with Bitmaps.
```cs
var np = new NDArray<int[][]>();
np.ARange(h, w, 3);
np.ReShape(h, w, 3);
for (int i = 0; i < dataLen; i++)
{
    var x = i % w;
    var y = i / w;
    var o = (y * stride + x * 4);
    np.Data[y][x] = new int[] { data[o + 3], data[o + 2], data[o + 1] };
}
```
At least now, all the data is in a numpy int[][][] array.
![debug1](https://user-images.githubusercontent.com/343840/47795081-c8dcad00-dcef-11e8-8add-7968e935fc82.png)

I have no idea how this works with any of the other operations.

**Comment by Oceania2018** [comment]
@fdncred Are you using the new NDArray?

**Comment by Oceania2018** [comment]
@dotChris90 I think we are discussing with dynamic dim and static dim.

![image](https://user-images.githubusercontent.com/1705364/47796420-4f928980-dcf2-11e8-91c0-a8a607682340.png)

You prefer static dim `NDArray<double[]>`, and I'm doing dynamic dim `NDArray<double>`.  I want to make sure we are on the same page.

![image](https://user-images.githubusercontent.com/1705364/47796717-e65f4600-dcf2-11e8-965f-55271cc47d86.png)

**Comment by Oceania2018** [comment]
@fdncred Pull the latest code. You can access element by `np[n,m,x]`

![image](https://user-images.githubusercontent.com/1705364/47797295-0e9b7480-dcf4-11e8-8913-2a5765468000.png)

**Comment by dotChris90** [comment]
Ah yes. This is true. U are right. In general spoken.

Dynamic can be a generic with value type. So double, float, int etc

Static can be a generic with Array type implementing IList. So double [] [], double [] or int [] [] [].

Got your point. I will try sth when home.

Sorry didn't get you before. But indeed that's nice.

**Comment by Oceania2018** [comment]
Ideally, we could mix `xarray` and `xtensor` in NumCpp. So NumSharp NDArray can handle double and double[][] very well. Hope to discuss further and show your implementation. :)

**Comment by dotChris90** [comment]
@Oceania2018 honestly this was exactly my idea. My first try failed. Seems you can not have 2 classes same name but different generic restrictions in C#

Damn. Honestly now I am thinking about implementing 2 different classes like in their approach and with a fast casting from dynamic to static so performance critical methods are executed there. Moreover we do not need to implement both classes methods lol 

But maybe find a better solution. I will investigate

**Comment by dotChris90** [comment]
Or since the generic restriction is really terrible for numeric. Maybe I add a generic type check in constructor. From the generic type we know if static or dynamic and that's it.

**Comment by fdncred** [comment]
@Oceania2018 Thanks for the tip on Reshape(). I'll play with that. I assume assignment works that same way, i.e. ```np[1,1,1] = 254;```

**Comment by fdncred** [comment]
I like this:
```cs
var np = new NDArray<byte>();
np.ARange(w * h * 3).ReShape(h, w, 3);
// 585.8108 ms to create, arange, reshape
for (int i = 0; i < dataLen; i++)
{
    var x = i % w;
    var y = i / w;
    var o = (y * stride + x * 4);
    np[y, x, 0] = data[o + 3];
    np[y, x, 1] = data[o + 2];
    np[y, x, 2] = data[o + 1];
}
//1141.8362 ms for assignment
```
Need to investigate parallelism to make it faster.

**Comment by dotChris90** [comment]
Ok even I am huge fan of static types. But have to admit it looks fantastic.

**Comment by fdncred** [comment]
Added .AsParallel() doesn't seem to improve this part.
```cs
public static NDArray<int> ARange(this NDArray<int> np, int stop, int start = 0, int step = 1)
{
    int index = 0;

    np.Data = Enumerable.Range(start, stop - start)
                        .Where(x => index++ % step == 0)
                        .AsParallel()
                        .ToArray();

    np.Shape = new List<int>() { stop };

    return np;
}
```
But parallel definitely helps in the assignment.
```cs
var sw = Stopwatch.StartNew();
var np = new NDArray<byte>();
np.ARange(w * h * 3).ReShape(h, w, 3);
sw.Stop();
var timer1 = sw.Elapsed.TotalMilliseconds;
// timer1 = 614.3598
sw.Restart();
//for (int i = 0; i < dataLen; i++)
System.Threading.Tasks.Parallel.For(0, dataLen, i =>
{
    var x = i % w;
    var y = i / w;
    var o = (y * stride + x * 4);
    np[y, x, 0] = data[o + 3];
    np[y, x, 1] = data[o + 2];
    np[y, x, 2] = data[o + 1];
});
sw.Stop();
var timer2 = sw.Elapsed.TotalMilliseconds;
// timer2 = 419.9634
```

**Comment by dotChris90** [comment]
@Oceania2018 @fdncred until now did not find a perfect way to handle both means in QuantStack Tensor and Array classes in one NDArray. So I will start to restructure our legacy NDArray into a NDTensor . After this finally our unit Tests shall stop failing .... :D

**Comment by dotChris90** [comment]
in later release we should try to combine both. :D

**Comment by Oceania2018** [comment]
@dotChris90 great plan. I'll do the Dynamic array NDArray, you will do the NDTensor. Appreciate.

**Comment by Oceania2018** [comment]
@dotChris90 interesting NumCpp `xt::xtensor<double, 2>({{3, 4}, {5, 6}})`, we might copy the constructor by defining your NDTensor as `public NDTensor<T>(int dim)`. Depends on your thought.

**Comment by dotChris90** [comment]
Yes. Indeed. Unfortunately I found out that the C# generics are less powerful than C++ templates. 

I tried something like 

public class NDTensor<T> 
{
    public T Data {get;set;}
}

Here I get problems when indexing. When indexing compiler tells me that it is unable to deduce the element type of matrix .... 

so have to do like 

public class NDTensor <TElement,T>
{
    public T Data {get;set;}

    public TElement this[int idx]
    { .... }
}

but this looks very ugly. 
Maybe I check out if we should also do the 1 dimensional Array strategy for all Tensors (matrix, array, tensor, ...) lets see ...

**Comment by Oceania2018** [comment]
We might focus on dynamic array NDArray before deciding how to code static dimension array.

**Comment by dotChris90** [comment]
Since I can't find good solutions I agree.

I will start by rewriting the NDArray operations like plus, minus, multiplication etc.
If you don't mind. :)

**Comment by Oceania2018** [comment]
No, please. I’ve commented some testing code. You can restore them back.

**Comment by dotChris90** [comment]
@Oceania2018 
Ok will do tomorrow. Today just did some performance tests. 

Even I did not like the idea of 1D array for Matrix, Vector, Tensor and so on, but! I see some benefits now from this. 

The element wise operations (element wise + or - ) are 1.5 - 2 times faster in 1D Array than in jagged array. I tested it for a 1000 x 1000 matrix (which would be a 1000000 1D Array). 

I will try it for matrix multiplications also. 

@fdncred @Oceania2018 
Just one question for curiosity. Does Numpy have sth. like 3D, 4D multiplication? Until now I just know 2D multiplication.

**Comment by fdncred** [comment]
@dotChris90 I'm no python expert but I do know that python does math on 3-dimensional arrays. Such as this [link ](https://docs.scipy.org/doc/numpy-1.13.0/reference/arrays.ndarray.html#calculation)shows.

**Comment by dotChris90** [comment]
@fdncred interesting. Thanks. I think this higher dimensions operations is not our goal but can keep in mind lol

**Comment by Oceania2018** [comment]
Glad to hear that, NumPy definitely do N-dimension multiplication. Same thing for us due to 1 dimension internally. You just need to loop the 1 dim array.

**Comment by dotChris90** [comment]
hello guys - seems Matrix Multiplikation is also faster with a 1D array I am very impressed. Ok I will be very happy tomorrow to rewrite the operations. It is awesome. Now I think - if we carefully implement the methods and operations - we can be faster than the C# higher dimension jagged arrays.

**Comment by Oceania2018** [comment]
After you compared performance benchmarks, could you also post the screenshots or leave the testing code?

**Comment by dotChris90** [comment]
U suggest to include them here at unit test or want a special folder benchmark on top level beside src and test?

Was the only reason why I didn't pushed.

Since until now just some basics tests done. I suggest to simply add a file to unit tests. And there are the benchmark I did until now.

Plus send images here. Do tomorrow morning before restructure operations.

**Comment by Oceania2018** [comment]
I prefer creat a new console project to adopt benchmark code.

**Comment by dotChris90** [comment]
Ok. Benchmark test which location?

/benchmark or at /test/benchmark?

**Comment by Oceania2018** [comment]
already create a new project at /test/benchmark
![image](https://user-images.githubusercontent.com/1705364/47914593-2a2a8a80-de6e-11e8-8104-946b66520e6c.png)

**Comment by dotChris90** [comment]
Ok my bad lol

**Comment by dotChris90** [comment]
Ok - I think we end this topic. 

@Oceania2018 @fdncred 
I was wrong - the .NET based multidimensional arrays  and jagged arrays are worse in performance and general use. I would create an issue which contains the "implementation of method which computes a jagged array or multidimensional array from the 1D array in storage". This I just want to do for a better integrate with other .NET libraries. 

@Oceania2018 @fdncred 
If you do not mind - I will close this issue. The quantstack 1D idea is awesome and generic with a super performance.

**Comment by fdncred** [comment]
I'm ready to close it.

**Comment by dotChris90** [comment]
Let's close it :)
